src/model/bloodvesseljunction.hpp

```diff
@@ -0,0 +1,256 @@
+   * @param y Current solution
+   */
+  void update_solution(ALGEBRA::SparseSystem<T> &system,
+                       Eigen::Matrix<T, Eigen::Dynamic, 1> &y);
+
+  /**
+   * @brief Number of triplets of element
+   *
+   * Number of triplets that the element contributes to the global system
+   * (relevant for sparse memory reservation)
+   */
+  std::map<std::string, int> num_triplets = {
+      {"F", 0},
+      {"E", 0},
+      {"D", 0},
+  };
+
+  /**
+   * @brief Get number of triplets of element
+   *
+   * Number of triplets that the element contributes to the global system
+   * (relevant for sparse memory reservation)
+   */
+  std::map<std::string, int> get_num_triplets();
+
+ private:
+  std::vector<BloodVessel<T> *> blood_vessels;
+  Parameters params;
+  unsigned int num_outlets;
+};
+
+template <typename T>
+BloodVesselJunction<T>::BloodVesselJunction(std::vector<T> R, std::vector<T> C,
+                                            std::vector<T> L,
+                                            std::vector<T> stenosis_coefficient,
+                                            std::string name)
+    : Block<T>(name) {
+  this->name = name;
+  this->params.R = R;
+  this->params.C = C;
+  this->params.L = L;
+  this->params.stenosis_coefficient = stenosis_coefficient;
+}
+
+template <typename T>
+BloodVesselJunction<T>::~BloodVesselJunction() {}
+
+template <typename T>
+void BloodVesselJunction<T>::setup_dofs(DOFHandler &dofhandler) {
+  // Set number of equations of a junction block based on number of
+  // inlets/outlets. Must be set before calling parent constructor
+  num_outlets = this->outlet_nodes.size();
+  std::list<std::string> internal_var_names;
+  for (size_t i = 0; i < num_outlets; i++) {
+    internal_var_names.push_back("flow_" + std::to_string(i));
+  }
+  Block<T>::setup_dofs_(dofhandler, 1, internal_var_names);
+  for (size_t i = 0; i < num_outlets; i++) {
+    blood_vessels.push_back(new BloodVessel<T>(
+        params.R[i], params.C[i], params.L[i], params.stenosis_coefficient[i],
+        this->name + "_bv" + std::to_string(i)));
+    blood_vessels[i]->inlet_nodes.push_back(this->inlet_nodes[0]);
+    blood_vessels[i]->outlet_nodes.push_back(this->outlet_nodes[i]);
+    blood_vessels[i]->setup_dofs(dofhandler);
+    blood_vessels[i]->global_var_ids[1] =
+        this->global_var_ids.end()[i - num_outlets];
+  }
+  num_triplets = {
+      {"F", 10 * num_outlets + num_outlets + 1},
+      {"E", 2 * num_outlets},
+      {"D", 2 * num_outlets},
+  };
+}
+
+template <typename T>
+void BloodVesselJunction<T>::update_constant(ALGEBRA::SparseSystem<T> &system) {
+  for (auto bv : blood_vessels) {
+    bv->update_constant(system);
+  }
+  // Mass conservation
+  system.F.coeffRef(this->global_eqn_ids[0], this->global_var_ids[1]) = 1.0;
+  for (size_t i = 0; i < num_outlets; i++) {
+    system.F.coeffRef(this->global_eqn_ids[0],
+                      this->global_var_ids.end()[-(i + 1)]) = -1.0;
```

Comment:
@richterjakob, you have to judge if it's worth it: Would it make sense to move mass conservation into a junction base class? Physically, we would expect this from every junction, regardless of how we model the pressure change.

From this, we could derive several specific junctions (that we already have): "normal" junction that preserves static pressure, resistive junction, BloodVessel junction.

This structure could make it slightly easier and less redundant (but more confusing) to add new junctions, like preserving total pressure or advanced models like @natalia-rubio is working on.

Reply:
The mass conservation here is not the same as in the normal junction class. For the normal junction, mass is conserved between inlets and outlets of the junction. This mass conservation here is a little hacky way to enable reusing the BloodVessel class. It ensures that the flow into the junction is correctly distributed to the **inlets** of the internal BloodVessel elements. I know this is a bit confusing and I didn't find a way to describe it better. Let me know, what you think. Should I add more documentation on that?

Reply:
Ah, I see your indexing is different. I think that's fine. I don't think it makes sense to change all indexing for the sake of inheritance.